### ISSUE TITLE
Add ability to manually track MMR and/or percent until next badge

### ISSUE THREAD

**Issue by dovchinnikov** [odota/web#1875]
**Is your feature request related to a problem? Please describe.**
Many players still track MMR with spreadsheets and it's inconvenient.

**Describe the solution you'd like**
Add an ability to manually enter the MMR after some match so MMR changes would be displayable in the table like in old Dota interface with +/- near each match. It should not be publicly available, i.e. private use only. It could be a separate tab in place of #1843.

This is what my table looks like:
<img width="630" alt="image" src="https://user-images.githubusercontent.com/5519549/46799051-6467a700-cd5c-11e8-9ae5-cd1984a7fd2f.png">

I cannot view the match details right away unless I insert a link in each row, I have to manually copy formula to compute the delta, and no other sites has such a feature.

What do you think?

**Comment by mobyte0** [comment]
As far as I know and off the top of my head, the only solution to this is with a client-side application for each person that wishes to track this information.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by howardchung** [comment]
Yeah, I'm not sure how many people would use this if you had to input all the data manually.  It wouldn't be much more beneficial than the spreadsheet approach, especially since we can't provide a match link either unless the user types in the match ID.

**Comment by dovchinnikov** [comment]
> especially since we can't provide a match link either unless the user types in the match ID

Wait, you have all the match IDs already, and I see this feature as a separate column with input fields in this view (or in the another separate tab):
<img width="1055" alt="screen shot 2018-10-23 at 13 43 43" src="https://user-images.githubusercontent.com/5519549/47389569-f57a3e80-d6c9-11e8-95f6-ed3330706539.png">

This data may be used to build a graph like the old one when MMR was public.

**Comment by howardchung** [comment]
The problem is that we can't link the match IDs with the changes in MMR.

**Comment by dovchinnikov** [comment]
@howardchung why? Sorry, I still don't get it.

**Comment by dovchinnikov** [comment]
> we can't check MMR anymore

Because Valve closed this API, right?

**Comment by dovchinnikov** [comment]
That's why I came here :) 
Since it's not possible to get MMR via official APIs, could you please add an input field near each ranked match, so the user could **manually** enter new MMR?

**Comment by howardchung** [comment]
Without any data verification I'm not sure we shouldn't store this data server-side since we won't know whether it's accurate.

Maybe it could be done client side and saved in localStorage but at that point it's not much more useful than a spreadsheet anyway?

**Comment by dovchinnikov** [comment]
> Without any data verification I'm not sure we shouldn't store this data server-side since we won't know whether it's accurate. 

Yes. That's why I've proposed that it should not be publicly available. 

> Maybe it could be done client side and saved in localStorage but at that point it's not much more useful than a spreadsheet anyway?

It will be useful since maintaining the spreadsheet is so much work. Spreadsheet cannot automatically:
- add a new row after match is finished, user has to add it manually after each match;
- insert match id after match is finished, user has to manually insert find it in client and insert it into the spreadsheet (I've given up on match IDs in my example);
- insert date and time (I'm adding date and ignoring the time for the sake of reducing the manual work to maintain the spreadsheet);
- compute MMR deltas, unless user is experienced with spreadsheet functions;
- highlight MMR deltas in green/red;
- track solo/party (I'm doing this is a single spreadsheet and it's not quite convenient).

All of the above is already available on the site, and the only missing piece is ability to manually enter the MMR value after each match.

These values should be visible to the user only, so it would make no sense to insert whatever MMR he wants. I.e. if users wants some stats at the end of the season, then opendota gives the ability to view them. The only validation is required that the value is a number > 0 and less than something reasonable.

I have multiple accounts (who doesn't? :)), and if the user won't lose all values after re-login, then local storage is ok.

> I suspect the usage of this feature would be very low since it requires significant manual effort.

I cannot comment on the usage of this feature, but I'd say it requires way less manual effort than spreadsheet approach.

**Comment by pacollins** [comment]
Chiming in here - I think it could be useful, but would clutter the screen unless there was a separate page for the analytics side of it and purely have the user type the MMR on the match page.

**Comment by ndeakin** [comment]
I'm interested in adding support for this: now that MMR change is fixed based on match conditions, no manual effort is required beyond entering a single reference point.

It would be possible to infer back to the last MMR changes and forward at least until another major MMR update (potentially beyond depending on the update).

One open question in my mind is, should this be private or public? Perhaps up to the user? The main reason for being private is that a player could easily just lie about their reference point, so this isn't necessarily "useful" data to others. On the flip-side, I can imagine wanting to share progress with friends. So perhaps it defaults to private, but an be made public?

Without having dug into the internals of the old page yet, it's not entirely clear to me if this should be in it's own separate view, or can reuse the old one.

Please let me know if you have thoughts!

**Comment by howardchung** [comment]
We can start with it public and then add restrictions on it as needed. I think it's fine as long we label the data clearly as self-reported.

On the server-side the changes are:
* New table to store self-reported MMR
* new POST API endpoint to accept MMR inputs
* add self-reported MMR to an existing endpoint or or existing endpoints
* add task to insert match workflow to update MMR values after ranked matches

Once we have the data we can think about how to display it. I guess we're interested both in a "current" MMR value and in MMR values after each match played (history?)

**Comment by ndeakin** [comment]
Thanks for the overview of the server side changes needed, and sounds good about leaving it public with statement that it's user-reported. I'll start looking into the details and may bug you in the discord.

Regarding current vs history, it's definitely accurate to say that my goal is to show current and history - the history is the big motivation for me personally, since the number alone is easy enough to grab in client. History on the other hand, isn't really tracked or shown anywhere any longer to my knowledge.
